packages/grafana-ui/src/components/GraphNG/GraphNG.tsx

```diff
@@ -20,137 +19,119 @@ export const FIXED_UNIT = '__fixed';
 }
 
 class UnthemedGraphNG extends React.Component<GraphNGProps, GraphNGState> {
   constructor(props: GraphNGProps) {
     super(props);
-
-    pluginLog('GraphNG', false, 'constructor, data aligment');
-    const alignedData = preparePlotFrame(props.data, {
-      x: fieldMatchers.get(FieldMatcherID.firstTimeField).get({}),
-      y: fieldMatchers.get(FieldMatcherID.numeric).get({}),
-    });
-
-    if (!alignedData) {
-      return;
-    }
-
-    this.state = {
-      alignedDataFrame: alignedData,
-      data: preparePlotData(alignedData),
-      config: preparePlotConfigBuilder(alignedData, props.theme, this.getTimeRange, this.getTimeZone),
-    };
+    this.state = this.prepState(props);
   }
 
-  componentDidUpdate(prevProps: GraphNGProps) {
-    const { theme, structureRev, data } = this.props;
-    let shouldConfigUpdate = false;
-    let stateUpdate = {} as GraphNGState;
-
-    if (this.state.config === undefined || this.props.timeZone !== prevProps.timeZone) {
-      shouldConfigUpdate = true;
-    }
+  getTimeRange = () => this.props.timeRange;
 
-    if (data !== prevProps.data) {
-      pluginLog('GraphNG', false, 'data changed');
-      const hasStructureChanged = structureRev !== prevProps.structureRev || !structureRev;
+  prepState(props: GraphNGProps, withConfig = true) {
+    let state: GraphNGState = null as any;
 
-      if (hasStructureChanged) {
-        pluginLog('GraphNG', false, 'schema changed');
-      }
+    const { frames, fields } = props;
 
-      pluginLog('GraphNG', false, 'componentDidUpdate, data aligment');
-      const alignedData = preparePlotFrame(data, {
+    const alignedFrame = preparePlotFrame(
+      frames,
+      fields || {
         x: fieldMatchers.get(FieldMatcherID.firstTimeField).get({}),
         y: fieldMatchers.get(FieldMatcherID.numeric).get({}),
-      });
-
-      if (!alignedData) {
-        return;
       }
+    );
 
-      stateUpdate = {
-        alignedDataFrame: alignedData,
-        data: preparePlotData(alignedData),
+    if (alignedFrame) {
+      state = {
+        alignedFrame,
+        alignedData: preparePlotData(alignedFrame, [FieldType.number]),
       };
 
-      if (shouldConfigUpdate || hasStructureChanged) {
-        pluginLog('GraphNG', false, 'updating config');
-        const builder = preparePlotConfigBuilder(alignedData, theme, this.getTimeRange, this.getTimeZone);
-        stateUpdate = { ...stateUpdate, config: builder };
+      if (withConfig) {
+        state.config = props.prepConfig(alignedFrame, this.getTimeRange);
       }
     }
 
-    if (Object.keys(stateUpdate).length > 0) {
-      this.setState(stateUpdate);
-    }
+    return state;
   }
 
-  getTimeRange = () => {
-    return this.props.timeRange;
-  };
+  componentDidUpdate(prevProps: GraphNGProps) {
+    const { frames, structureRev, timeZone, propsToDiff } = this.props;
 
-  getTimeZone = () => {
-    return this.props.timeZone;
-  };
+    if (frames !== prevProps.frames) {
```

Comment:
@ryantxu @dprokop , in panel Edit mode, this condition does not get triggered in Timeline when changing options, but _is_ triggered in TimeSeries when changing options. not sure why...